pkg/controllers/configobservation/customroutersecret/observe_custom_router_secret.go

```diff
@@ -0,0 +1,135 @@
+		value, found, err := unstructured.NestedFieldCopy(existing, fields...)
+		if !found {
+			continue
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+		err = unstructured.SetNestedField(previous, value, fields...)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return previous, errs
+}
+
+func (c *customRouterObserver) observe(genericlisters configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (ret map[string]interface{}, _ []error) {
+	listers := genericlisters.(configobservation.Listers)
+	errs := []error{}
+
+	// pick the correct resource sync function
+	resourceSync := listers.ResourceSyncer().SyncSecret
+	if _, ok := c.resourceType.(corev1.ConfigMap); ok {
+		resourceSync = listers.ResourceSyncer().SyncConfigMap
+	}
+
+	previouslyObservedConfig, errs := extractPreviouslyObservedConfig(existingConfig, c.configPaths...)
+
+	ingress, err := listers.IngressLister.Get("cluster")
+	// if something went wrong, keep the previously observed config and resources
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedComponentRoutes, observedResources, err := c.componentRouteSecret(listers, ingress)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedConfig := map[string]interface{}{}
+	if err := unstructured.SetNestedField(observedConfig, observedComponentRoutes, c.configPaths[0]...); err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	errs = append(errs, syncObservedResources(resourceSync, observedResources)...)
+	return observedConfig, errs
+}
+
+// resourceSyncFunc syncs a resource from the source location to the destination location.
+type resourceSyncFunc func(destination, source resourcesynccontroller.ResourceLocation) error
+
+// syncActionRules rules define source resource names indexed by destination resource names.
+// Empty value means to delete the destination.
+type syncActionRules map[string]string
+
+// syncObservedResources copies or deletes resources, sources in GlobalUserSpecifiedConfigNamespace and destinations in OperatorNamespace namespace.
+// Errors are collected, i.e. it's not failing on first error.
+func syncObservedResources(syncResource resourceSyncFunc, syncRules syncActionRules) []error {
+	var errs []error
+	for to, from := range syncRules {
+		var source resourcesynccontroller.ResourceLocation
+		if len(from) > 0 {
+			source = resourcesynccontroller.ResourceLocation{Namespace: "openshift-config", Name: from}
+		}
+		// if 'from' is empty, then it means we want to delete
+		destination := resourcesynccontroller.ResourceLocation{Namespace: "openshift-authentication", Name: to}
+		if err := syncResource(destination, source); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+func (c *customRouterObserver) componentRouteSecret(lister configobservation.Listers, ingress *configv1.Ingress) ([]interface{}, syncActionRules, error) {
```

Comment:
you never appear to use error in this function

Reply:
I'll remove the error.